crates/daphne-server/src/storage_proxy_connection/kv/mod.rs

```diff
@@ -252,28 +253,64 @@ impl<'h> Kv<'h> {
     }
 
     #[tracing::instrument(
-        name = "kv_put",
+        name = "kv_delete",
         skip_all,
         fields(key, prefix = std::any::type_name::<P>()),
     )]
-    pub async fn put<P>(&self, key: &P::Key, value: P::Value) -> Result<(), Error>
+    pub async fn put_internal<P>(
+        &self,
+        key: &P::Key,
+        value: P::Value,
+        expiration: Option<Time>,
+    ) -> Result<(), Error>
     where
         P: KvPrefix,
         P::Key: std::fmt::Debug,
+        P::Value: Serialize,
     {
         let key = Self::to_key::<P>(key);
+        // let expiration_str = expiration.map_or_else(String::new, |time| time.to_string());
         tracing::debug!(key, "PUT");
-        self.http
+
+        let mut request = self
+            .http
             .post(self.config.url.join(&key).unwrap())
             .bearer_auth(&self.config.auth_token)
-            .body(serde_json::to_vec(&value).unwrap())
-            .send()
-            .await?
-            .error_for_status()?;
+            .body(serde_json::to_vec(&value).unwrap());
+
+        if let Some(expiration) = expiration {
+            request = request.header("X-Expiration", expiration.to_string());
```

Comment:
This is a header intended for a public API. To minimize the chance of someone accidentally overwriting this in a request, I think we should make it more specific. Perhaps something like
```suggestion
            request = request.header("x-daphne-storage-proxy-kv-put-expiration", expiration.to_string());
```
Also, I believe lower case is preferred. (Compliant HTTP libraries will always convert to lower case.)